docs/develop/compiler.rst

```diff
@@ -0,0 +1,417 @@
+
+            // Check if any implicitly declared variables should be closed over
+            for (size_t i = 0; i < s->id_info_len; ++i) {
+                id_info_t *id = &s->id_info[i];
+                if (id->kind == ID_INFO_KIND_GLOBAL_IMPLICIT) {
+                    scope_check_to_close_over(s, id);
+                }
+            }
+        }
+   ..
+   }
+
+Other computations regarding scopes and identifiers are computed too.
+At this point, the number of labels that will be required in the emitted code is also determined and set.
+
+Second and third passes
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The second and third passes involve computing the code size and emitting the ``inline assembler code`` for
+the different architectures:
+
+.. code-block:: c
+   
+   // compile pass 2 and 3
+   #if MICROPY_EMIT_NATIVE
+   emit_t *emit_native = NULL;
+   #endif
+   for (scope_t *s = comp->scope_head; s != NULL && comp->compile_error == MP_OBJ_NULL; s = s->next) {
+        #if MICROPY_EMIT_INLINE_ASM
+        if (s->emit_options == MP_EMIT_OPT_ASM) {
+            // inline assembly
+            if (comp->emit_inline_asm == NULL) {
+                comp->emit_inline_asm = ASM_EMITTER(new)(max_num_labels);
+            }
+            comp->emit = NULL;
+            comp->emit_inline_asm_method_table = ASM_EMITTER_TABLE;
+            compile_scope_inline_asm(comp, s, MP_PASS_CODE_SIZE);
+            #if MICROPY_EMIT_INLINE_XTENSA
+            // Xtensa requires an extra pass to compute size of l32r const table
+            // TODO this can be improved by calculating it during SCOPE pass
+            // but that requires some other structural changes to the asm emitters
+            #if MICROPY_DYNAMIC_COMPILER
+            if (mp_dynamic_compiler.native_arch == MP_NATIVE_ARCH_XTENSA)
+            #endif
+            {
+                compile_scope_inline_asm(comp, s, MP_PASS_CODE_SIZE);
+            }
+            #endif
+            if (comp->compile_error == MP_OBJ_NULL) {
+                compile_scope_inline_asm(comp, s, MP_PASS_EMIT);
+            }
+        } else
+
+The inline assembler code comprises assembly instructions in a Python function.
+See `the inline assembler tutorial 
+<https://docs.micropython.org/en/latest/pyboard/tutorial/assembler.html#pyboard-tutorial-assembler>`_ 
+for more details.
+
+Fourth, fifth and sixth passes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The other two passes compute the stack and code size, while the last pass emits the final code:
+
+.. code-block:: c
+   
+   compile_scope(comp, s, MP_PASS_STACK_SIZE);
+
+   if (comp->compile_error == MP_OBJ_NULL) {
+        compile_scope(comp, s, MP_PASS_CODE_SIZE);
+   }
+
+   if (comp->compile_error == MP_OBJ_NULL) {
+    compile_scope(comp, s, MP_PASS_EMIT);
+   }
+
+Before these passes, there is a selection for the type of code to be emitted which can either be native or
+bytecode.
+
+.. code-block:: c
+
+   switch (s->emit_options) {
+
+        #if MICROPY_EMIT_NATIVE
```

Comment:
as above, I think it'd be clearer to the reader if these #if/#endif lines were removed (but the lines between them kept)